cumulus/pallets/xcmp-queue/src/lib.rs

```diff
@@ -791,22 +788,46 @@ impl<T: Config> XcmpMessageHandler for Pallet<T> {
 							},
 						}
 					},
-				XcmpMessageFormat::ConcatenatedVersionedXcm =>
+				XcmpMessageFormat::ConcatenatedVersionedXcm => {
+					// We need to know if the current message is the first on the current XCMP page
+					// for weight metering accuracy.
+					let mut is_first_xcm_on_page = true;
 					while !data.is_empty() {
 						let Ok(xcm) = Self::take_first_concatenated_xcm(&mut data, &mut meter)
 						else {
 							defensive!("HRMP inbound decode stream broke; page will be dropped.",);
 							break
 						};
 
-						if let Err(()) = Self::enqueue_xcmp_message(sender, xcm, &mut meter) {
+						// For simplicity, we consider that each new XCMP page results in a new
+						// message queue page. This is not always true, but it's a good enough
```

Comment:
It could be that every XCMP message is also a new MQ page? Or are there some runtime integrity tests in place to check the relation between XCMP and MQ page size?

Reply:
I don't think there are any integrity tests related to this. An XCMP message can result in no new MQ page, but it can also result in multiple MQ pages. I just did a rough estimate here. But I think I can improve the accuracy of tracking MQ pages and also I plan to do batching since it seems much faster. Will get back with some improvements soon.

Reply:
I opened a follow-up PR for this: https://github.com/paritytech/polkadot-sdk/pull/8021